src/inmanta/deploy/tasks.py

```diff
@@ -306,3 +328,61 @@ async def execute(self, task_manager: "scheduler.TaskManager", agent: str, reaso
 
         get_fact_report = await my_executor.get_facts(executor_resource_details)
         await task_manager.fact_refresh_done(get_fact_report)
+
+
+class ModuleLoadingException(Exception):
+    """
+    This exception is raised when some Inmanta modules couldn't be loaded on a given agent.
+    """
+
+    def __init__(self, agent_name: str, failed_modules: FailedInmantaModules) -> None:
+        """
+        :param agent_name: Name of the agent for which module loading was unsuccessful
+        :param failed_modules: Data for all module loading errors as a nested map of
+            inmanta module name -> python module name -> Exception.
+        """
+        self.failed_modules = failed_modules
+        self.agent_name = agent_name
+
+    def _format_module_loading_errors(self) -> str:
+        """
+        Helper method to display module loading failures.
+        """
+        formatted_module_loading_errors = ""
+        N_FAILURES = sum(len(v) for v in self.failed_modules.values())
+        failure_index = 1
+
+        for _, failed_modules_data in self.failed_modules.items():
+            for python_module, exception in failed_modules_data.items():
+                formatted_module_loading_errors += f"Error {failure_index}/{N_FAILURES}:\n"
+                formatted_module_loading_errors += f"In module {python_module}:\n"
+                formatted_module_loading_errors += str(exception)
+                failure_index += 1
+
+        return formatted_module_loading_errors
+
+    def create_log_line_for_failed_modules(self, level: int = logging.ERROR, *, verbose_message: bool = False) -> data.LogLine:
+        """
+        Helper method to convert this Exception into a LogLine.
+
+        :param level: The log level for the resulting LogLine
+        :param verbose_message: Whether to include the full formatted error output in the LogLine message.
+            When displayed on the webconsole, the full formatted error output will be displayed in its own section
+            regardless of this flag's value.
+
+        """
+        message = "Agent %s failed to load the following modules: %s." % (
+            self.agent_name,
+            ", ".join(self.failed_modules.keys()),
+        )
+
+        formatted_module_loading_errors = self._format_module_loading_errors()
+
+        if verbose_message:
+            message += f"\n{formatted_module_loading_errors}"
+        return data.LogLine.log(
+            level=level,
+            msg=message,
+            timestamp=None,
+            errors=formatted_module_loading_errors,
+        )
```

Comment:
Q1: Is this an acceptable usage of the LogLine class ? i.e.
- _direct_ substitutions in `message` instead of named substitutions with kwargs
- an "extra" kwarg `errors` that is not meant to be substituted.

Looks like this:

![image](https://github.com/user-attachments/assets/8bbd9515-e4c4-4f8d-99aa-82c8f90975c9)